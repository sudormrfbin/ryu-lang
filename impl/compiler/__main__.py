from compiler.repl import repl

repl()

# with open("input.lang", "r") as f:
#     inp = f.read()

# output = compile(inp)
# print(output)
