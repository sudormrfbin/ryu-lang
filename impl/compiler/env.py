from typing import Any

from .errors import InternalCompilerError


class RuntimeEnvironment:
    values: dict[str, Any]

    def __init__(self):
        self.values = {}

    def define(self, name: str, value: Any):
        # shadowing is allowed
        self.values[name] = value

    def get(self, name: str) -> Any:
        try:
            return self.values[name]
        except KeyError:
            raise InternalCompilerError(f"Variable {name} not defined at runtime")
