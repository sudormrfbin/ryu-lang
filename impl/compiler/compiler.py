from typing import Any

from compiler.env import RuntimeEnvironment, TypeEnvironment

from compiler import builtins, errors, langvalues
from compiler.parser import parse, parse_tree_to_ast

BUILTIN_FUNCTIONS: list[type[langvalues.BuiltinFunction]] = [
    builtins.SumFunction,
    builtins.ArrayLengthFunction,
    builtins.StringLengthFunction,
    builtins.ArrayAppend,
]


def get_default_environs() -> tuple[TypeEnvironment, RuntimeEnvironment]:
    type_env = TypeEnvironment()
    runtime_env = RuntimeEnvironment()

    for fn in BUILTIN_FUNCTIONS:
        type_env.define(fn.TYPE.function_name, fn.TYPE)
        runtime_env.define(fn.TYPE.function_name, fn())

    return (type_env, runtime_env)


def run(source: str, type_env: TypeEnvironment, runtime_env: RuntimeEnvironment) -> Any:
    try:
        return _run(source, type_env, runtime_env)
    except errors.CompilerError as err:
        err.report(source)


def _run(
    source: str, type_env: TypeEnvironment, runtime_env: RuntimeEnvironment
) -> Any:
    tree = parse(source)
    ast = parse_tree_to_ast(tree)

    ast.typecheck(type_env)

    return ast.eval(runtime_env)
