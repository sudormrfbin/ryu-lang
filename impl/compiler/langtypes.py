from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Optional, Sequence, TypeAlias
from typing_extensions import override


if TYPE_CHECKING:
    from compiler.env import TypeEnvironment
    from compiler.errors import Span


class Type:
    @property
    def name(self) -> str:
        return type(self).__name__

    @classmethod
    def from_str(cls, ident: str, env: TypeEnvironment) -> Optional["Type"]:
        if ident in PRIMITIVE_TYPES:
            return PRIMITIVE_TYPES[ident]

        ty = env.get(ident)
        if isinstance(ty, UserDefinedType):
            return ty

        return None

    @classmethod
    def from_str_with_generics(
        cls,
        ident: str,
        generics: Type,
        env: TypeEnvironment,
    ) -> Optional["Type"]:
        if ident == "array":
            return Array(generics)

        return None


class Placeholder(Type):
    pass


class PrimitiveType(Type):
    pass


class UserDefinedType(Type):
    pass


class Bool(PrimitiveType):
    pass


class Int(PrimitiveType):
    pass


class String(PrimitiveType):
    pass


class Block(Type):
    pass


@dataclass
class ReturnBlock(Block):
    return_type: Type
    return_stmt_span: Span


@dataclass
class Struct(UserDefinedType):
    struct_name: str
    members: Members

    @dataclass
    class Members(PrimitiveType):
        types: dict[str, Type]

        def __len__(self) -> int:
            return len(self.types)


@dataclass
class Enum(UserDefinedType):
    enum_name: str
    members: list[EnumVariantSimple | EnumVariantTuple]
    span: Span

    @property
    @override
    def name(self) -> str:
        return self.enum_name

    def variant_from_str(
        self, name: str
    ) -> EnumVariantSimple | EnumVariantTuple | None:
        for mem in self.members:
            if mem.name == name:
                return mem
        return None


@dataclass
class EnumVariantSimple:
    name: str


@dataclass
class EnumVariantTuple:
    name: str
    inner: Type


EnumVariants: TypeAlias = EnumVariantSimple | EnumVariantTuple


@dataclass
class Array(PrimitiveType):
    ty: Type

    @property
    @override
    def name(self) -> str:
        return f"Array<{self.ty.name}>"


@dataclass
class UntypedArray(PrimitiveType):
    pass


@dataclass
class Function(PrimitiveType):
    function_name: str
    arguments: Params
    return_type: Type


@dataclass
class Params(PrimitiveType):
    types: list[Type]

    def __len__(self) -> int:
        return len(self.types)


BOOL = Bool()
INT = Int()
STRING = String()
BLOCK = Block()
PLACEHOLDER = Placeholder()


PRIMITIVE_TYPES: dict[str, Type] = {
    "bool": BOOL,
    "int": INT,
    "string": STRING,
}


def resolve_blocks_type(types: Sequence[Type]) -> Block:
    resolved = BLOCK
    for ty in types:
        match (resolved, ty):
            case (Block(), ReturnBlock()):
                resolved = ty
            case (ReturnBlock(), ReturnBlock()):
                if resolved.return_type != ty.return_type:
                    raise  # TODO different return types
            case _:
                pass

    return resolved
