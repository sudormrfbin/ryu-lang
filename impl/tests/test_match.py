from typing import Any
from compiler.env import RuntimeEnvironment, TypeEnvironment
from compiler.parser import parse, parse_tree_to_ast
from compiler.ast import (
    Assignment,
    BoolLiteral,
    CaseLadder,
    CaseStmt,
    IntLiteral,
    MatchStmt,
    StatementBlock,
    StatementList,
    UnaryOp,
    VariableDeclaration,
)
from compiler.langtypes import INT, Block, Bool, Int
from tests.utils import docstring_source, docstring_source_with_snapshot


@docstring_source
def test_match_case_bool(source: str):
    """
    let result = -1

    match true {
        case true { result = 1 }
        case false { result = 0 }
    }
    """
    ast = parse_tree_to_ast(parse(source))
    assert ast.to_dict() == {
        StatementList: {
            "stmts": [
                {
                    VariableDeclaration: {
                        "ident": "result",
                        "rvalue": {
                            UnaryOp: {"op": "-", "operand": {IntLiteral: {"value": 1}}}
                        },
                    }
                },
                {
                    MatchStmt: {
                        "expr": {BoolLiteral: {"value": True}},
                        "cases": {
                            CaseLadder: {
                                "cases": [
                                    {
                                        CaseStmt: {
                                            "pattern": {BoolLiteral: {"value": True}},
                                            "block": {
                                                StatementBlock: {
                                                    "stmts": [
                                                        {
                                                            Assignment: {
                                                                "lvalue": "result",
                                                                "rvalue": {
                                                                    IntLiteral: {
                                                                        "value": 1
                                                                    }
                                                                },
                                                            }
                                                        }
                                                    ]
                                                }
                                            },
                                        }
                                    },
                                    {
                                        CaseStmt: {
                                            "pattern": {BoolLiteral: {"value": False}},
                                            "block": {
                                                StatementBlock: {
                                                    "stmts": [
                                                        {
                                                            Assignment: {
                                                                "lvalue": "result",
                                                                "rvalue": {
                                                                    IntLiteral: {
                                                                        "value": 0
                                                                    }
                                                                },
                                                            }
                                                        }
                                                    ]
                                                }
                                            },
                                        }
                                    },
                                ]
                            }
                        },
                    }
                },
            ]
        }
    }

    type_env = TypeEnvironment()
    ast.typecheck(type_env)
    assert ast.to_type_dict() == {
        StatementList: Block,
        "stmts": [
            {
                VariableDeclaration: Int,
                "rvalue": {UnaryOp: Int, "operand": {IntLiteral: Int}},
            },
            {
                MatchStmt: Block,
                "expr": {BoolLiteral: Bool},
                "cases": {
                    CaseLadder: Block,
                    "cases": [
                        {
                            CaseStmt: Bool,
                            "pattern": {BoolLiteral: Bool},
                            "block": {
                                StatementBlock: Block,
                                "stmts": [
                                    {Assignment: Int, "rvalue": {IntLiteral: Int}}
                                ],
                            },
                        },
                        {
                            CaseStmt: Bool,
                            "pattern": {BoolLiteral: Bool},
                            "block": {
                                StatementBlock: Block,
                                "stmts": [
                                    {Assignment: Int, "rvalue": {IntLiteral: Int}}
                                ],
                            },
                        },
                    ],
                },
            },
        ],
    }

    assert type_env.get("result") == INT

    env = RuntimeEnvironment()
    ast.eval(env)
    assert env.get("result") == 1


@docstring_source_with_snapshot
def test_enum_pattern_match_wildcard(source: str, snapshot: Any):
    """
    enum Langs {
        Malayalam
        English
        Japanese
    }

    fn is_eng(lang: Langs) -> bool {
        match lang {
            case Langs::English { return true }
            case _ { return false }
        }
    }

    let with_eng = is_eng(Langs::English)
    let with_mal = is_eng(Langs::Malayalam)
    let with_jp = is_eng(Langs::Japanese)
    """
    ast = parse_tree_to_ast(parse(source))
    assert ast.to_dict() == snapshot(name="ast")

    type_env = TypeEnvironment()
    ast.typecheck(type_env)
    assert ast.to_type_dict() == snapshot(name="typed-ast")

    env = RuntimeEnvironment()
    ast.eval(env)
    assert env.get("with_eng") is True
    assert env.get("with_mal") is False
    assert env.get("with_jp") is False


@docstring_source_with_snapshot
def test_match_array(source: str, snapshot: Any):
    """
    fn len(arr: array<int>) -> int {
        match arr {
            case [1] { return 1 }
            case [1, 2] { return 2 }
            case [1, 2, 3] { return 3 }
            case _ { return -1 }
        }
    }

    let one = len([1])
    let two = len([1, 2])
    let three = len([1, 2, 3])
    let more = len([1, 2, 3, 4])
    """
    ast = parse_tree_to_ast(parse(source))
    assert ast.to_dict() == snapshot(name="ast")

    type_env = TypeEnvironment()
    ast.typecheck(type_env)
    assert ast.to_type_dict() == snapshot(name="typed-ast")

    env = RuntimeEnvironment()
    ast.eval(env)
    assert env.get("one") == 1
    assert env.get("two") == 2
    assert env.get("three") == 3
    assert env.get("more") == -1


@docstring_source_with_snapshot
def test_match_array_empty_case(source: str, snapshot: Any):
    """
    fn len(arr: array<int>) -> int {
        match arr {
            case [] { return 0 }
            case _ { return -1 }
        }
    }

    let zero = len(<int>[])
    let one = len([1])
    let two = len([1, 2])
    let three = len([1, 2, 3])
    let more = len([1, 2, 3, 4])
    """
    ast = parse_tree_to_ast(parse(source))
    assert ast.to_dict() == snapshot(name="ast")

    type_env = TypeEnvironment()
    ast.typecheck(type_env)
    assert ast.to_type_dict() == snapshot(name="typed-ast")

    env = RuntimeEnvironment()
    ast.eval(env)
    assert env.get("zero") == 0
    assert env.get("one") == -1
    assert env.get("two") == -1
    assert env.get("three") == -1
    assert env.get("more") == -1
