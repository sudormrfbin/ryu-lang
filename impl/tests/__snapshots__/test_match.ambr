# serializer version: 1
# name: test_enum_pattern_match_wildcard[ast]
  dict({
    <class 'compiler.ast.StatementList'>: dict({
      'stmts': list([
        dict({
          <class 'compiler.ast.EnumStmt'>: dict({
            'members': dict({
              <class 'compiler.ast.EnumMembers'>: dict({
                'members': list([
                  dict({
                    <class 'compiler.ast.EnumMember'>: dict({
                      'name': Token('IDENTIFIER', 'Malayalam'),
                    }),
                  }),
                  dict({
                    <class 'compiler.ast.EnumMember'>: dict({
                      'name': Token('IDENTIFIER', 'English'),
                    }),
                  }),
                  dict({
                    <class 'compiler.ast.EnumMember'>: dict({
                      'name': Token('IDENTIFIER', 'Japanese'),
                    }),
                  }),
                ]),
              }),
            }),
            'name': Token('IDENTIFIER', 'Langs'),
          }),
        }),
        dict({
          <class 'compiler.ast.FunctionDefinition'>: dict({
            'args': dict({
              <class 'compiler.ast.FunctionParams'>: dict({
                'args': list([
                  dict({
                    <class 'compiler.ast.FunctionParam'>: dict({
                      'arg_type': dict({
                        <class 'compiler.ast.TypeAnnotation'>: dict({
                          'ty': Token('IDENTIFIER', 'Langs'),
                        }),
                      }),
                      'name': Token('IDENTIFIER', 'lang'),
                    }),
                  }),
                ]),
              }),
            }),
            'body': dict({
              <class 'compiler.ast.StatementBlock'>: dict({
                'stmts': list([
                  dict({
                    <class 'compiler.ast.MatchStmt'>: dict({
                      'cases': dict({
                        <class 'compiler.ast.CaseLadder'>: dict({
                          'cases': list([
                            dict({
                              <class 'compiler.ast.CaseStmt'>: dict({
                                'block': dict({
                                  <class 'compiler.ast.StatementBlock'>: dict({
                                    'stmts': list([
                                      dict({
                                        <class 'compiler.ast.ReturnStmt'>: dict({
                                          'return_value': dict({
                                            <class 'compiler.ast.BoolLiteral'>: dict({
                                              'value': True,
                                            }),
                                          }),
                                        }),
                                      }),
                                    ]),
                                  }),
                                }),
                                'pattern': dict({
                                  <class 'compiler.ast.EnumLiteral'>: dict({
                                    'enum_type': Token('IDENTIFIER', 'Langs'),
                                    'variant': Token('IDENTIFIER', 'English'),
                                  }),
                                }),
                              }),
                            }),
                            dict({
                              <class 'compiler.ast.CaseStmt'>: dict({
                                'block': dict({
                                  <class 'compiler.ast.StatementBlock'>: dict({
                                    'stmts': list([
                                      dict({
                                        <class 'compiler.ast.ReturnStmt'>: dict({
                                          'return_value': dict({
                                            <class 'compiler.ast.BoolLiteral'>: dict({
                                              'value': False,
                                            }),
                                          }),
                                        }),
                                      }),
                                    ]),
                                  }),
                                }),
                                'pattern': dict({
                                  <class 'compiler.ast.WildcardPattern'>: dict({
                                  }),
                                }),
                              }),
                            }),
                          ]),
                        }),
                      }),
                      'expr': dict({
                        <class 'compiler.ast.Variable'>: dict({
                          'value': Token('IDENTIFIER', 'lang'),
                        }),
                      }),
                    }),
                  }),
                ]),
              }),
            }),
            'name': Token('IDENTIFIER', 'is_eng'),
            'return_type': dict({
              <class 'compiler.ast.TypeAnnotation'>: dict({
                'ty': Token('IDENTIFIER', 'bool'),
              }),
            }),
          }),
        }),
        dict({
          <class 'compiler.ast.VariableDeclaration'>: dict({
            'ident': Token('IDENTIFIER', 'with_eng'),
            'rvalue': dict({
              <class 'compiler.ast.FunctionCall'>: dict({
                'args': dict({
                  <class 'compiler.ast.FunctionArgs'>: dict({
                    'args': list([
                      dict({
                        <class 'compiler.ast.EnumLiteral'>: dict({
                          'enum_type': Token('IDENTIFIER', 'Langs'),
                          'variant': Token('IDENTIFIER', 'English'),
                        }),
                      }),
                    ]),
                  }),
                }),
                'callee': dict({
                  <class 'compiler.ast.Variable'>: dict({
                    'value': Token('IDENTIFIER', 'is_eng'),
                  }),
                }),
              }),
            }),
          }),
        }),
        dict({
          <class 'compiler.ast.VariableDeclaration'>: dict({
            'ident': Token('IDENTIFIER', 'with_mal'),
            'rvalue': dict({
              <class 'compiler.ast.FunctionCall'>: dict({
                'args': dict({
                  <class 'compiler.ast.FunctionArgs'>: dict({
                    'args': list([
                      dict({
                        <class 'compiler.ast.EnumLiteral'>: dict({
                          'enum_type': Token('IDENTIFIER', 'Langs'),
                          'variant': Token('IDENTIFIER', 'Malayalam'),
                        }),
                      }),
                    ]),
                  }),
                }),
                'callee': dict({
                  <class 'compiler.ast.Variable'>: dict({
                    'value': Token('IDENTIFIER', 'is_eng'),
                  }),
                }),
              }),
            }),
          }),
        }),
        dict({
          <class 'compiler.ast.VariableDeclaration'>: dict({
            'ident': Token('IDENTIFIER', 'with_jp'),
            'rvalue': dict({
              <class 'compiler.ast.FunctionCall'>: dict({
                'args': dict({
                  <class 'compiler.ast.FunctionArgs'>: dict({
                    'args': list([
                      dict({
                        <class 'compiler.ast.EnumLiteral'>: dict({
                          'enum_type': Token('IDENTIFIER', 'Langs'),
                          'variant': Token('IDENTIFIER', 'Japanese'),
                        }),
                      }),
                    ]),
                  }),
                }),
                'callee': dict({
                  <class 'compiler.ast.Variable'>: dict({
                    'value': Token('IDENTIFIER', 'is_eng'),
                  }),
                }),
              }),
            }),
          }),
        }),
      ]),
    }),
  })
# ---
# name: test_enum_pattern_match_wildcard[typed-ast]
  dict({
    'stmts': list([
      dict({
        'members': dict({
          'members': list([
            dict({
              <class 'compiler.ast.EnumMember'>: <class 'compiler.langtypes.Type'>,
            }),
            dict({
              <class 'compiler.ast.EnumMember'>: <class 'compiler.langtypes.Type'>,
            }),
            dict({
              <class 'compiler.ast.EnumMember'>: <class 'compiler.langtypes.Type'>,
            }),
          ]),
          <class 'compiler.ast.EnumMembers'>: <class 'compiler.langtypes.Block'>,
        }),
        <class 'compiler.ast.EnumStmt'>: <class 'compiler.langtypes.Enum'>,
      }),
      dict({
        'args': dict({
          'args': list([
            dict({
              'arg_type': dict({
                <class 'compiler.ast.TypeAnnotation'>: <class 'compiler.langtypes.Enum'>,
              }),
              <class 'compiler.ast.FunctionParam'>: <class 'compiler.langtypes.Enum'>,
            }),
          ]),
          <class 'compiler.ast.FunctionParams'>: <class 'compiler.langtypes.Params'>,
        }),
        'body': dict({
          'stmts': list([
            dict({
              'cases': dict({
                'cases': list([
                  dict({
                    'block': dict({
                      'stmts': list([
                        dict({
                          'return_value': dict({
                            <class 'compiler.ast.BoolLiteral'>: <class 'compiler.langtypes.Bool'>,
                          }),
                          <class 'compiler.ast.ReturnStmt'>: <class 'compiler.langtypes.ReturnBlock'>,
                        }),
                      ]),
                      <class 'compiler.ast.StatementBlock'>: <class 'compiler.langtypes.ReturnBlock'>,
                    }),
                    'pattern': dict({
                      <class 'compiler.ast.EnumLiteral'>: <class 'compiler.langtypes.Enum'>,
                    }),
                    <class 'compiler.ast.CaseStmt'>: <class 'compiler.langtypes.Enum'>,
                  }),
                  dict({
                    'block': dict({
                      'stmts': list([
                        dict({
                          'return_value': dict({
                            <class 'compiler.ast.BoolLiteral'>: <class 'compiler.langtypes.Bool'>,
                          }),
                          <class 'compiler.ast.ReturnStmt'>: <class 'compiler.langtypes.ReturnBlock'>,
                        }),
                      ]),
                      <class 'compiler.ast.StatementBlock'>: <class 'compiler.langtypes.ReturnBlock'>,
                    }),
                    'pattern': dict({
                      <class 'compiler.ast.WildcardPattern'>: <class 'compiler.langtypes.Placeholder'>,
                    }),
                    <class 'compiler.ast.CaseStmt'>: <class 'compiler.langtypes.Placeholder'>,
                  }),
                ]),
                <class 'compiler.ast.CaseLadder'>: <class 'compiler.langtypes.ReturnBlock'>,
              }),
              'expr': dict({
                <class 'compiler.ast.Variable'>: <class 'compiler.langtypes.Enum'>,
              }),
              <class 'compiler.ast.MatchStmt'>: <class 'compiler.langtypes.ReturnBlock'>,
            }),
          ]),
          <class 'compiler.ast.StatementBlock'>: <class 'compiler.langtypes.ReturnBlock'>,
        }),
        'return_type': dict({
          <class 'compiler.ast.TypeAnnotation'>: <class 'compiler.langtypes.Bool'>,
        }),
        <class 'compiler.ast.FunctionDefinition'>: <class 'compiler.langtypes.Function'>,
      }),
      dict({
        'rvalue': dict({
          'args': dict({
            'args': list([
              dict({
                <class 'compiler.ast.EnumLiteral'>: <class 'compiler.langtypes.Enum'>,
              }),
            ]),
            <class 'compiler.ast.FunctionArgs'>: <class 'compiler.langtypes.Params'>,
          }),
          'callee': dict({
            <class 'compiler.ast.Variable'>: <class 'compiler.langtypes.Function'>,
          }),
          <class 'compiler.ast.FunctionCall'>: <class 'compiler.langtypes.Bool'>,
        }),
        <class 'compiler.ast.VariableDeclaration'>: <class 'compiler.langtypes.Bool'>,
      }),
      dict({
        'rvalue': dict({
          'args': dict({
            'args': list([
              dict({
                <class 'compiler.ast.EnumLiteral'>: <class 'compiler.langtypes.Enum'>,
              }),
            ]),
            <class 'compiler.ast.FunctionArgs'>: <class 'compiler.langtypes.Params'>,
          }),
          'callee': dict({
            <class 'compiler.ast.Variable'>: <class 'compiler.langtypes.Function'>,
          }),
          <class 'compiler.ast.FunctionCall'>: <class 'compiler.langtypes.Bool'>,
        }),
        <class 'compiler.ast.VariableDeclaration'>: <class 'compiler.langtypes.Bool'>,
      }),
      dict({
        'rvalue': dict({
          'args': dict({
            'args': list([
              dict({
                <class 'compiler.ast.EnumLiteral'>: <class 'compiler.langtypes.Enum'>,
              }),
            ]),
            <class 'compiler.ast.FunctionArgs'>: <class 'compiler.langtypes.Params'>,
          }),
          'callee': dict({
            <class 'compiler.ast.Variable'>: <class 'compiler.langtypes.Function'>,
          }),
          <class 'compiler.ast.FunctionCall'>: <class 'compiler.langtypes.Bool'>,
        }),
        <class 'compiler.ast.VariableDeclaration'>: <class 'compiler.langtypes.Bool'>,
      }),
    ]),
    <class 'compiler.ast.StatementList'>: <class 'compiler.langtypes.Block'>,
  })
# ---
# name: test_match_array[ast]
  dict({
    <class 'compiler.ast.StatementList'>: dict({
      'stmts': list([
        dict({
          <class 'compiler.ast.VariableDeclaration'>: dict({
            'ident': Token('IDENTIFIER', 'len'),
            'rvalue': dict({
              <class 'compiler.ast.IntLiteral'>: dict({
                'value': 0,
              }),
            }),
          }),
        }),
        dict({
          <class 'compiler.ast.MatchStmt'>: dict({
            'cases': dict({
              <class 'compiler.ast.CaseLadder'>: dict({
                'cases': list([
                  dict({
                    <class 'compiler.ast.CaseStmt'>: dict({
                      'block': dict({
                        <class 'compiler.ast.StatementBlock'>: dict({
                          'stmts': list([
                            dict({
                              <class 'compiler.ast.Assignment'>: dict({
                                'lvalue': Token('IDENTIFIER', 'len'),
                                'rvalue': dict({
                                  <class 'compiler.ast.IntLiteral'>: dict({
                                    'value': 1,
                                  }),
                                }),
                              }),
                            }),
                          ]),
                        }),
                      }),
                      'pattern': dict({
                        <class 'compiler.ast.ArrayPattern'>: dict({
                          'elements': list([
                            dict({
                              <class 'compiler.ast.ArrayPatternElement'>: dict({
                                'literal': dict({
                                  <class 'compiler.ast.IntLiteral'>: dict({
                                    'value': 1,
                                  }),
                                }),
                              }),
                            }),
                          ]),
                        }),
                      }),
                    }),
                  }),
                  dict({
                    <class 'compiler.ast.CaseStmt'>: dict({
                      'block': dict({
                        <class 'compiler.ast.StatementBlock'>: dict({
                          'stmts': list([
                            dict({
                              <class 'compiler.ast.Assignment'>: dict({
                                'lvalue': Token('IDENTIFIER', 'len'),
                                'rvalue': dict({
                                  <class 'compiler.ast.IntLiteral'>: dict({
                                    'value': 2,
                                  }),
                                }),
                              }),
                            }),
                          ]),
                        }),
                      }),
                      'pattern': dict({
                        <class 'compiler.ast.ArrayPattern'>: dict({
                          'elements': list([
                            dict({
                              <class 'compiler.ast.ArrayPatternElement'>: dict({
                                'literal': dict({
                                  <class 'compiler.ast.IntLiteral'>: dict({
                                    'value': 1,
                                  }),
                                }),
                              }),
                            }),
                            dict({
                              <class 'compiler.ast.ArrayPatternElement'>: dict({
                                'literal': dict({
                                  <class 'compiler.ast.IntLiteral'>: dict({
                                    'value': 2,
                                  }),
                                }),
                              }),
                            }),
                          ]),
                        }),
                      }),
                    }),
                  }),
                  dict({
                    <class 'compiler.ast.CaseStmt'>: dict({
                      'block': dict({
                        <class 'compiler.ast.StatementBlock'>: dict({
                          'stmts': list([
                            dict({
                              <class 'compiler.ast.Assignment'>: dict({
                                'lvalue': Token('IDENTIFIER', 'len'),
                                'rvalue': dict({
                                  <class 'compiler.ast.IntLiteral'>: dict({
                                    'value': 3,
                                  }),
                                }),
                              }),
                            }),
                          ]),
                        }),
                      }),
                      'pattern': dict({
                        <class 'compiler.ast.ArrayPattern'>: dict({
                          'elements': list([
                            dict({
                              <class 'compiler.ast.ArrayPatternElement'>: dict({
                                'literal': dict({
                                  <class 'compiler.ast.IntLiteral'>: dict({
                                    'value': 1,
                                  }),
                                }),
                              }),
                            }),
                            dict({
                              <class 'compiler.ast.ArrayPatternElement'>: dict({
                                'literal': dict({
                                  <class 'compiler.ast.IntLiteral'>: dict({
                                    'value': 2,
                                  }),
                                }),
                              }),
                            }),
                            dict({
                              <class 'compiler.ast.ArrayPatternElement'>: dict({
                                'literal': dict({
                                  <class 'compiler.ast.IntLiteral'>: dict({
                                    'value': 3,
                                  }),
                                }),
                              }),
                            }),
                          ]),
                        }),
                      }),
                    }),
                  }),
                  dict({
                    <class 'compiler.ast.CaseStmt'>: dict({
                      'block': dict({
                        <class 'compiler.ast.StatementBlock'>: dict({
                          'stmts': list([
                            dict({
                              <class 'compiler.ast.Assignment'>: dict({
                                'lvalue': Token('IDENTIFIER', 'len'),
                                'rvalue': dict({
                                  <class 'compiler.ast.UnaryOp'>: dict({
                                    'op': Token('MINUS', '-'),
                                    'operand': dict({
                                      <class 'compiler.ast.IntLiteral'>: dict({
                                        'value': 1,
                                      }),
                                    }),
                                  }),
                                }),
                              }),
                            }),
                          ]),
                        }),
                      }),
                      'pattern': dict({
                        <class 'compiler.ast.WildcardPattern'>: dict({
                        }),
                      }),
                    }),
                  }),
                ]),
              }),
            }),
            'expr': dict({
              <class 'compiler.ast.ArrayLiteral'>: dict({
                'members': dict({
                  <class 'compiler.ast.ArrayElements'>: dict({
                    'members': list([
                      dict({
                        <class 'compiler.ast.ArrayElement'>: dict({
                          'element': dict({
                            <class 'compiler.ast.IntLiteral'>: dict({
                              'value': 1,
                            }),
                          }),
                        }),
                      }),
                      dict({
                        <class 'compiler.ast.ArrayElement'>: dict({
                          'element': dict({
                            <class 'compiler.ast.IntLiteral'>: dict({
                              'value': 2,
                            }),
                          }),
                        }),
                      }),
                    ]),
                  }),
                }),
              }),
            }),
          }),
        }),
      ]),
    }),
  })
# ---
# name: test_match_array[typed-ast]
  dict({
    'stmts': list([
      dict({
        'rvalue': dict({
          <class 'compiler.ast.IntLiteral'>: <class 'compiler.langtypes.Int'>,
        }),
        <class 'compiler.ast.VariableDeclaration'>: <class 'compiler.langtypes.Int'>,
      }),
      dict({
        'cases': dict({
          'cases': list([
            dict({
              'block': dict({
                'stmts': list([
                  dict({
                    'rvalue': dict({
                      <class 'compiler.ast.IntLiteral'>: <class 'compiler.langtypes.Int'>,
                    }),
                    <class 'compiler.ast.Assignment'>: <class 'compiler.langtypes.Int'>,
                  }),
                ]),
                <class 'compiler.ast.StatementBlock'>: <class 'compiler.langtypes.Block'>,
              }),
              'pattern': dict({
                'elements': list([
                  dict({
                    'literal': dict({
                      <class 'compiler.ast.IntLiteral'>: <class 'compiler.langtypes.Int'>,
                    }),
                    <class 'compiler.ast.ArrayPatternElement'>: <class 'compiler.langtypes.Int'>,
                  }),
                ]),
                <class 'compiler.ast.ArrayPattern'>: <class 'compiler.langtypes.Array'>,
              }),
              <class 'compiler.ast.CaseStmt'>: <class 'compiler.langtypes.Array'>,
            }),
            dict({
              'block': dict({
                'stmts': list([
                  dict({
                    'rvalue': dict({
                      <class 'compiler.ast.IntLiteral'>: <class 'compiler.langtypes.Int'>,
                    }),
                    <class 'compiler.ast.Assignment'>: <class 'compiler.langtypes.Int'>,
                  }),
                ]),
                <class 'compiler.ast.StatementBlock'>: <class 'compiler.langtypes.Block'>,
              }),
              'pattern': dict({
                'elements': list([
                  dict({
                    'literal': dict({
                      <class 'compiler.ast.IntLiteral'>: <class 'compiler.langtypes.Int'>,
                    }),
                    <class 'compiler.ast.ArrayPatternElement'>: <class 'compiler.langtypes.Int'>,
                  }),
                  dict({
                    'literal': dict({
                      <class 'compiler.ast.IntLiteral'>: <class 'compiler.langtypes.Int'>,
                    }),
                    <class 'compiler.ast.ArrayPatternElement'>: <class 'compiler.langtypes.Int'>,
                  }),
                ]),
                <class 'compiler.ast.ArrayPattern'>: <class 'compiler.langtypes.Array'>,
              }),
              <class 'compiler.ast.CaseStmt'>: <class 'compiler.langtypes.Array'>,
            }),
            dict({
              'block': dict({
                'stmts': list([
                  dict({
                    'rvalue': dict({
                      <class 'compiler.ast.IntLiteral'>: <class 'compiler.langtypes.Int'>,
                    }),
                    <class 'compiler.ast.Assignment'>: <class 'compiler.langtypes.Int'>,
                  }),
                ]),
                <class 'compiler.ast.StatementBlock'>: <class 'compiler.langtypes.Block'>,
              }),
              'pattern': dict({
                'elements': list([
                  dict({
                    'literal': dict({
                      <class 'compiler.ast.IntLiteral'>: <class 'compiler.langtypes.Int'>,
                    }),
                    <class 'compiler.ast.ArrayPatternElement'>: <class 'compiler.langtypes.Int'>,
                  }),
                  dict({
                    'literal': dict({
                      <class 'compiler.ast.IntLiteral'>: <class 'compiler.langtypes.Int'>,
                    }),
                    <class 'compiler.ast.ArrayPatternElement'>: <class 'compiler.langtypes.Int'>,
                  }),
                  dict({
                    'literal': dict({
                      <class 'compiler.ast.IntLiteral'>: <class 'compiler.langtypes.Int'>,
                    }),
                    <class 'compiler.ast.ArrayPatternElement'>: <class 'compiler.langtypes.Int'>,
                  }),
                ]),
                <class 'compiler.ast.ArrayPattern'>: <class 'compiler.langtypes.Array'>,
              }),
              <class 'compiler.ast.CaseStmt'>: <class 'compiler.langtypes.Array'>,
            }),
            dict({
              'block': dict({
                'stmts': list([
                  dict({
                    'rvalue': dict({
                      'operand': dict({
                        <class 'compiler.ast.IntLiteral'>: <class 'compiler.langtypes.Int'>,
                      }),
                      <class 'compiler.ast.UnaryOp'>: <class 'compiler.langtypes.Int'>,
                    }),
                    <class 'compiler.ast.Assignment'>: <class 'compiler.langtypes.Int'>,
                  }),
                ]),
                <class 'compiler.ast.StatementBlock'>: <class 'compiler.langtypes.Block'>,
              }),
              'pattern': dict({
                <class 'compiler.ast.WildcardPattern'>: <class 'compiler.langtypes.Placeholder'>,
              }),
              <class 'compiler.ast.CaseStmt'>: <class 'compiler.langtypes.Placeholder'>,
            }),
          ]),
          <class 'compiler.ast.CaseLadder'>: <class 'compiler.langtypes.Block'>,
        }),
        'expr': dict({
          'members': dict({
            'members': list([
              dict({
                'element': dict({
                  <class 'compiler.ast.IntLiteral'>: <class 'compiler.langtypes.Int'>,
                }),
                <class 'compiler.ast.ArrayElement'>: <class 'compiler.langtypes.Int'>,
              }),
              dict({
                'element': dict({
                  <class 'compiler.ast.IntLiteral'>: <class 'compiler.langtypes.Int'>,
                }),
                <class 'compiler.ast.ArrayElement'>: <class 'compiler.langtypes.Int'>,
              }),
            ]),
            <class 'compiler.ast.ArrayElements'>: <class 'compiler.langtypes.Array'>,
          }),
          <class 'compiler.ast.ArrayLiteral'>: <class 'compiler.langtypes.Array'>,
        }),
        <class 'compiler.ast.MatchStmt'>: <class 'compiler.langtypes.Block'>,
      }),
    ]),
    <class 'compiler.ast.StatementList'>: <class 'compiler.langtypes.Block'>,
  })
# ---
